File: imstargg-frontend/src/model/PlayerBattle.ts
import {BattlePlayer} from "@/model/BattlePlayer";
import {BattleType} from "@/model/enums/BattleType";
import {BattleResult} from "@/model/enums/BattleResult";
import {BattleMode} from "@/model/enums/BattleMode";
import {PlayerBattleEvent} from "@/model/PlayerBattleEvent";

export interface PlayerBattle {
  battleTime: Date;
  event: PlayerBattleEvent;
  mode: BattleMode;
  type: BattleType;
  result?: BattleResult;
  duration?: number;
  rank?: number;
  trophyChange?: number;
  starPlayerTag: string | null;
  teams: BattlePlayer[][];
}

export function playerBattleMyTeam(battle: PlayerBattle, tag: string): BattlePlayer[] {
  return battle.teams.find((team) => team.some((player) => player.tag === tag)) ?? [];
}

export function playerBattleMe(battle: PlayerBattle, tag: string): BattlePlayer[] {
  return battle.teams.flat().filter((player) => player.tag === tag);
}

File: core/core-domain/src/main/java/com/imstargg/core/domain/player/PlayerBattle.java
package com.imstargg.core.domain.player;

import com.imstargg.core.domain.BrawlStarsTag;
import com.imstargg.core.enums.BattleMode;
import com.imstargg.core.enums.BattleResult;
import com.imstargg.core.enums.BattleType;
import jakarta.annotation.Nullable;

import java.time.OffsetDateTime;
import java.util.List;

public record PlayerBattle(
        OffsetDateTime battleTime,
        PlayerBattleEvent event,
        BattleMode mode,
        BattleType type,
        @Nullable BattleResult result,
        @Nullable Integer duration,
        @Nullable Integer rank,
        @Nullable Integer trophyChange,
        @Nullable BrawlStarsTag starPlayerTag,
        List<List<BattlePlayer>> teams
) {
}
